### PR TITLE
PERF: Series.fillna for pyarrow-backed dtypes 

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.0.0.rst` file if fixing a bug or adding a new feature.

Perf improvement in `Series.fillna` for pyarrow-backed dtypes.

 ```
       before           after         ratio
     [86f18282]       [34319599]
     <main>           <arrow-ea-fillna>
-        98.6±1ms       18.7±0.4ms     0.19  series_methods.Fillna.time_fillna('string[pyarrow]', 'backfill')
-        99.2±1ms       17.6±0.2ms     0.18  series_methods.Fillna.time_fillna('string[pyarrow]', 'pad')
-       1.86±0.1s      3.86±0.03ms     0.00  series_methods.Fillna.time_fillna('int64[pyarrow]', 'backfill')
-      1.89±0.07s      3.01±0.03ms     0.00  series_methods.Fillna.time_fillna('int64[pyarrow]', 'pad')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

### CODE REVIEW

File: pandas/core/arrays/arrow/array.py [hunk 1]
```diff
@@ -521,6 +525,44 @@ def dropna(self: ArrowExtensionArrayT) -> ArrowExtensionArrayT:
...
+    def fillna(
+        self: ArrowExtensionArrayT,
+        value: object | ArrayLike | None = None,
+        method: FillnaOptions | None = None,
+        limit: int | None = None,
+    ) -> ArrowExtensionArrayT:
+
+        value, method = validate_fillna_kwargs(value, method)
+
+        if isinstance(self._data.type, pa.DurationType):
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:538]: Could this be put in a try except instead? So if this is implemented in a future pyarrow version, this still won't be going down the `super` path

Comment (jbrockmendel) [reply pandas/core/arrays/arrow/array.py:538]: maybe with a TODO with condition upon which it can be removed

Comment (lukemanley) [reply pandas/core/arrays/arrow/array.py]: Updated to use try except along with a future TODO.

File: pandas/core/arrays/arrow/array.py [hunk 2]
```diff
@@ -521,6 +525,48 @@ def dropna(self: ArrowExtensionArrayT) -> ArrowExtensionArrayT:
...
+                    val = pa.scalar(value, type=self._data.type, from_pandas=True)
+                return type(self)(pc.fill_null(self._data, fill_value=val))
+
+            elif method == "pad" and limit is None and not pa_version_under7p0:
+                return type(self)(pc.fill_null_forward(self._data))
+
+            elif method == "backfill" and limit is None and not pa_version_under7p0:
+                return type(self)(pc.fill_null_backward(self._data))
+
+        except (pa.ArrowInvalid, pa.ArrowTypeError):
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:559]: Can these `try/excepts` be more tightly scoped to where these exceptions are expected (guessing the `pa.array` and `pc.fill_null*` calls

Comment (lukemanley) [reply pandas/core/arrays/arrow/array.py]: yep, good point. updated to more tightly scope

File: pandas/core/arrays/arrow/array.py [hunk 3]
```diff
@@ -521,6 +525,57 @@ def dropna(self: ArrowExtensionArrayT) -> ArrowExtensionArrayT:
...
+            if isinstance(value, (pa.Scalar, pa.Array, pa.ChunkedArray)):
+                return value
+            if is_array_like(value):
+                func = pa.array
+            else:
+                func = pa.scalar
+            try:
+                value = func(value, type=pa_type, from_pandas=True)
+            except pa.ArrowTypeError:
+                raise TypeError(f"Invalid value '{str(value)}' for dtype {dtype}")
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:558]: Nit: Could you `raise TypeError(...) from err` where `err = pa.ArrowTypeError`?

File: pandas/core/arrays/arrow/array.py [hunk 4]
```diff
@@ -521,6 +525,57 @@ def dropna(self: ArrowExtensionArrayT) -> ArrowExtensionArrayT:
...
+                    f" expected {len(self)}"
+                )
+
+        def convert_fill_value(value, pa_type, dtype):
+            if value is None:
+                return value
+            if isinstance(value, (pa.Scalar, pa.Array, pa.ChunkedArray)):
+                return value
+            if is_array_like(value):
+                func = pa.array
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:552]: Nit: Could you call this `pa_box` instead of `func`?

File: pandas/core/arrays/arrow/array.py [hunk 5]
```diff
@@ -521,6 +525,57 @@ def dropna(self: ArrowExtensionArrayT) -> ArrowExtensionArrayT:
...
+            try:
+                value = func(value, type=pa_type, from_pandas=True)
+            except pa.ArrowTypeError:
+                raise TypeError(f"Invalid value '{str(value)}' for dtype {dtype}")
+            return value
+
+        fill_value = convert_fill_value(value, self._data.type, self.dtype)
+
+        try:
+            if method is None and limit is None:
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:564]: Nit: Could you factor out the `limit is None` our of the try except?

File: pandas/core/arrays/arrow/array.py [hunk 6]
```diff
@@ -521,6 +525,57 @@ def dropna(self: ArrowExtensionArrayT) -> ArrowExtensionArrayT:
...
+                return type(self)(pc.fill_null(self._data, fill_value=fill_value))
+            elif method == "pad" and limit is None and not pa_version_under7p0:
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:566]: Given the use of `pa_version_under7p0`, we want to call `fallback_performancewarning` that a slower path (the super path) is being called

File: pandas/core/arrays/arrow/array.py [hunk 7]
```diff
@@ -521,6 +525,66 @@ def dropna(self: ArrowExtensionArrayT) -> ArrowExtensionArrayT:
...
+    ) -> ArrowExtensionArrayT:
+
+        value, method = validate_fillna_kwargs(value, method)
+
+        if limit is not None:
+            return super().fillna(value=value, method=method, limit=limit)
+
+        if method is not None and pa_version_under7p0:
+            # fill_null_{forward|backward} added in pyarrow 7.0
+            fallback_performancewarning(version="7")
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:543]: Could you make sure these are being checked with `tm.maybe_produces_warning` in the test suite (should show up in the pyarrow=6 build)? Ideally we want to catch all occurrences of these in this test suite

**Comment by mroeschke** [comment]
Looking good. Just a single comment

File: pandas/core/arrays/arrow/array.py (hunk 7 above)
Comment (lukemanley) [reply pandas/core/arrays/arrow/array.py:543]: added. thanks for the reviews

**Comment by mroeschke** [comment]
A few other places where this is showing up otherwise LGTM 

```
pandas/tests/extension/test_string.py::TestMissing::test_fillna_no_op_returns_copy[True-pyarrow]
pandas/tests/extension/test_string.py::TestMissing::test_fillna_no_op_returns_copy[False-pyarrow]
  /home/runner/work/pandas/pandas/pandas/tests/extension/base/missing.py:93: PerformanceWarning: Falling back on a non-pyarrow code path which may decrease performance. Upgrade to pyarrow >=7 to possibly suppress this warning.
    result = data.fillna(method="backfill")

pandas/tests/extension/test_string.py::TestMissing::test_fillna_series_method[ffill-True-pyarrow]
pandas/tests/extension/test_string.py::TestMissing::test_fillna_series_method[ffill-False-pyarrow]
pandas/tests/extension/test_string.py::TestMissing::test_fillna_series_method[bfill-True-pyarrow]
pandas/tests/extension/test_string.py::TestMissing::test_fillna_series_method[bfill-False-pyarrow]
  /home/runner/work/pandas/pandas/pandas/tests/extension/base/missing.py:123: PerformanceWarning: Falling back on a non-pyarrow code path which may decrease performance. Upgrade to pyarrow >=7 to possibly suppress this warning.
    result = pd.Series(data_missing).fillna(method=fillna_method)
```

**Comment by lukemanley** [comment]
> A few other places where this is showing up otherwise LGTM

fixed, thx

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lukemanley

**Comment by lukemanley** [comment]
cc @mroeschke, @jbrockmendel 

It turns out `Block.fillna` does not dispatch to the underlying arrays in all cases including the simple case where method and limit are None. This highlights the impact of not dispatching for Arrow-backed arrays:

```
import pandas as pd
import pandas._testing as tm

N = 10**6
arr = pd.array(tm.rands_array(10, N), dtype="string[pyarrow]")
arr[::2] = None
ser = pd.Series(arr)

%timeit ser.fillna("foo")
60.5 ms ± 119 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)     # <- no dispatch

%timeit arr.fillna("foo")
4.89 ms ± 22.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)   # <- potential, if dispatched
``` 

It looks like the pattern might be to define `ArrowBackedExtensionBlock(ExtensionBlock)` with a `fillna` method that dispatches to `ArrowExtensionArray.fillna`.  However, the `Block.fillna` method is marked final. Any suggestions? Would it be problematic to remove the final attribute?

**Comment by mroeschke** [comment]
I imagine `ExtensionBlock` should generally define and just dispatch to to the array's `fillna` since (at least for the foreseeable future) they will be 1D?

**Comment by jbrockmendel** [comment]
> I imagine ExtensionBlock should generally define and just dispatch to to the array's fillna

That makes sense to me.  Would want to make sure we're not hurting perf in non-arrow cases.

**Pull request merged**